pyteal/ast/subroutine.py

```diff
@@ -461,53 +648,93 @@ def evaluateSubroutine(subroutine: SubroutineDefinition) -> SubroutineDeclaratio
-            loaded = argVar
+            argument_var = DynamicScratchVar(TealType.anytype)
+            loaded_var = argument_var
         elif param in subroutine.abi_args:
             internal_abi_var = subroutine.abi_args[param].new_instance()
-            argVar = internal_abi_var.stored_value
-            loaded = internal_abi_var
+            argument_var = internal_abi_var.stored_value
+            loaded_var = internal_abi_var
         else:
-            argVar = ScratchVar(TealType.anytype)
-            loaded = argVar.load()
+            argument_var = ScratchVar(TealType.anytype)
+            loaded_var = argument_var.load()
 
-        return argVar, loaded
+        return argument_var, loaded_var
+
+    if len(subroutine.output_kwarg) > 1:
+        raise TealInputError(
+            f"ABI keyword argument num: {len(subroutine.output_kwarg)}. "
+            f"Exceeding abi output keyword argument max number 1."
+        )
 
     args = subroutine.arguments()
-    argumentVars, loadedArgs = zip(*map(var_n_loaded, args)) if args else ([], [])
+    args = [arg for arg in args if arg not in subroutine.output_kwarg]
+
+    arg_vars: list[ScratchVar] = []
+    loaded_args: list[ScratchVar | Expr | abi.BaseType] = []
+    for arg in args:
+        arg_var, loaded_arg = var_n_loaded(arg)
+        arg_vars.append(arg_var)
+        loaded_args.append(loaded_arg)
+
+    abi_output_kwargs: dict[str, abi.BaseType] = {}
+    output_kwarg_info = OutputKwArgInfo.from_dict(subroutine.output_kwarg)
+    output_carrying_abi: Optional[abi.BaseType] = None
+
+    if output_kwarg_info:
+        output_carrying_abi = output_kwarg_info.abi_type.new_instance()
+        abi_output_kwargs[output_kwarg_info.name] = output_carrying_abi
 
     # Arg usage "B" supplied to build an AST from the user-defined PyTEAL function:
-    subroutineBody = subroutine.implementation(*loadedArgs)
+    subroutine_body = subroutine.implementation(*loaded_args, **abi_output_kwargs)
 
-    if not isinstance(subroutineBody, Expr):
+    if not isinstance(subroutine_body, Expr):
         raise TealInputError(
-            f"Subroutine function does not return a PyTeal expression. Got type {type(subroutineBody)}"
+            f"Subroutine function does not return a PyTeal expression. Got type {type(subroutine_body)}."
+        )
+    # if there is an output keyword argument for ABI, place the storing on the stack
+    if output_carrying_abi:
+        if subroutine_body.has_return():
+            raise TealInputError(
+                "ABI returning subroutine definition should have no return"
+            )
+        if subroutine_body.type_of() != TealType.none:
+            raise TealInputError(
+                f"ABI returning subroutine definition should evaluate to TealType.none, "
+                f"while evaluate to {subroutine_body.type_of()}."
+            )
+        subroutine_body = Seq(
+            subroutine_body, Return(output_carrying_abi.stored_value.load())
```

Comment:
Do you think we should worry about if a user never stores anything in the output variable? Or does PyTeal's built-in scratch slot load-before-store logic already prevent this?

In any case, do you think it should or shouldn't be allowed?

Reply:
This is a good point of thinking.

I do think it is not allowed to have output kwarg, and not storing anything inside. Essentially this will incur load before store error in ABI implementation.

While people can go down the path, and cause error from ABI (and current subroutine implementation), do you think we want to additionally check this and provide a specific error for defining output kwarg but not using it?

Reply:
Got it, as long as there is some sort of error, I think that's acceptable. I want to avoid reading from an uninitialized scratch slot, since that will produce a value that doesn't make sense, and it sounds like our current scratch slot load-before-store error is good enough.